RightCurly false-positives for multi-block continuation

https://checkstyle.org/config_blocks.html#RightCurly
https://checkstyle.org/property_types.html#rcurly `same` is default mode

```
/var/tmp $ javac Foo.java

/var/tmp $ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="RightCurly">
            <property name="tokens" value="CLASS_DEF"/>
            <property name="tokens" value="CTOR_DEF"/>
            <property name="tokens" value="METHOD_DEF"/>
        </module>
    </module>
</module>

/var/tmp $ cat Foo.java
class Foo {
    public Foo() {
        Thread t = new Thread(new Runnable() {
           @Override
           public void run() {
           } // <--
        });
    } // <--

    public void getId() {
    } // <--
} // <--

/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-8.17-all.jar -c config.xml Foo.java
Starting audit...
[ERROR] /var/tmp/Foo.java:6:12: '}' at column 12 should be on the same line as the next part of a multi-block statement (one that directly contains multiple blocks: if/else-if/else, do/while or try/catch/finally). [RightCurly]
[ERROR] /var/tmp/Foo.java:8:5: '}' at column 5 should be on the same line as the next part of a multi-block statement (one that directly contains multiple blocks: if/else-if/else, do/while or try/catch/finally). [RightCurly]
[ERROR] /var/tmp/Foo.java:11:5: '}' at column 5 should be on the same line as the next part of a multi-block statement (one that directly contains multiple blocks: if/else-if/else, do/while or try/catch/finally). [RightCurly]
[ERROR] /var/tmp/Foo.java:12:1: '}' at column 1 should be on the same line as the next part of a multi-block statement (one that directly contains multiple blocks: if/else-if/else, do/while or try/catch/finally). [RightCurly]
Audit done.
Checkstyle ends with 4 errors.
```

Expected:
No violation
Default mode is `same` https://checkstyle.org/property_types.html#rcurly

From doc:
> Same - 	Works like alone_or_singleline but the brace should be on the same line as the next part of a multi-block statement (one that directly contains multiple blocks: if/else-if/else or try/catch/finally). If no next part of a multi-block statement present, brace must be alone on line. It also allows single-line format of multi-block statements.

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java
@@ -222,10 +222,9 @@ public class RightCurlyCheck extends AbstractCheck {
                                                String targetSrcLine) {
         return bracePolicy == RightCurlyOption.ALONE
                     && shouldBeAloneOnLineWithAloneOption(details, targetSrcLine)
-                || bracePolicy == RightCurlyOption.ALONE_OR_SINGLELINE
-                    && shouldBeAloneOnLineWithAloneOrSinglelineOption(details, targetSrcLine)
-                || details.shouldCheckLastRcurly
-                    && details.rcurly.getLineNo() == details.nextToken.getLineNo();
+                || (bracePolicy == RightCurlyOption.ALONE_OR_SINGLELINE
+                    || details.shouldCheckLastRcurly)
+                    && shouldBeAloneOnLineWithNotAloneOption(details, targetSrcLine);
     }
 
     /**
@@ -240,17 +239,16 @@ public class RightCurlyCheck extends AbstractCheck {
     }
 
     /**
-     * Whether right curly should be alone on line when ALONE_OR_SINGLELINE option is used.
+     * Whether right curly should be alone on line when ALONE_OR_SINGLELINE or SAME option is used.
      * @param details details for validation.
      * @param targetSrcLine A string with contents of rcurly's line
      * @return true, if right curly should be alone on line
-     *         when ALONE_OR_SINGLELINE option is used.
+     *         when ALONE_OR_SINGLELINE or SAME option is used.
      */
-    private static boolean shouldBeAloneOnLineWithAloneOrSinglelineOption(Details details,
-                                                                          String targetSrcLine) {
-        return !isAloneOnLine(details, targetSrcLine)
-                && !isSingleLineBlock(details)
-                && !isEmptyBody(details.lcurly);
+    private static boolean shouldBeAloneOnLineWithNotAloneOption(Details details,
+                                                                 String targetSrcLine) {
+        return shouldBeAloneOnLineWithAloneOption(details, targetSrcLine)
+                && !isSingleLineBlock(details);
     }
 
     /**
@@ -315,32 +313,16 @@ public class RightCurlyCheck extends AbstractCheck {
     }
 
     /**
-     * Checks if definition body is empty.
-     * @param lcurly left curly.
-     * @return true if definition body is empty.
-     */
-    private static boolean isEmptyBody(DetailAST lcurly) {
-        boolean result = false;
-        if (lcurly.getParent().getType() == TokenTypes.OBJBLOCK) {
-            if (lcurly.getNextSibling().getType() == TokenTypes.RCURLY) {
-                result = true;
-            }
-        }
-        else if (lcurly.getFirstChild().getType() == TokenTypes.RCURLY) {
-            result = true;
-        }
-        return result;
-    }
-
-    /**
      * Checks if right curly has line break before.
      * @param rightCurly right curly token.
      * @return true, if right curly has line break before.
      */
     private static boolean hasLineBreakBefore(DetailAST rightCurly) {
-        final DetailAST previousToken = rightCurly.getPreviousSibling();
-        return previousToken == null
-                || rightCurly.getLineNo() != previousToken.getLineNo();
+        DetailAST previousToken = rightCurly.getPreviousSibling();
+        if (previousToken == null) {
+            previousToken = rightCurly.getParent();
+        }
+        return rightCurly.getLineNo() != previousToken.getLineNo();
     }
 
     /**
@@ -407,8 +389,7 @@ public class RightCurlyCheck extends AbstractCheck {
          * @return object containing all details to make a validation
          */
         private static Details getDetailsForTryCatchFinally(DetailAST ast) {
-            boolean shouldCheckLastRcurly = false;
-            final DetailAST rcurly;
+            final boolean shouldCheckLastRcurly;
             final DetailAST lcurly;
             DetailAST nextToken;
             final int tokenType = ast.getType();
@@ -420,28 +401,32 @@ public class RightCurlyCheck extends AbstractCheck {
                     lcurly = ast.getFirstChild();
                 }
                 nextToken = lcurly.getNextSibling();
-                rcurly = lcurly.getLastChild();
 
                 if (nextToken == null) {
                     shouldCheckLastRcurly = true;
                     nextToken = getNextToken(ast);
                 }
+                else {
+                    shouldCheckLastRcurly = false;
+                }
             }
             else if (tokenType == TokenTypes.LITERAL_CATCH) {
                 nextToken = ast.getNextSibling();
                 lcurly = ast.getLastChild();
-                rcurly = lcurly.getLastChild();
                 if (nextToken == null) {
                     shouldCheckLastRcurly = true;
                     nextToken = getNextToken(ast);
+                }
+                else {
+                    shouldCheckLastRcurly = false;
                 }
             }
             else {
                 shouldCheckLastRcurly = true;
                 nextToken = getNextToken(ast);
                 lcurly = ast.getFirstChild();
-                rcurly = lcurly.getLastChild();
             }
+            final DetailAST rcurly = lcurly.getLastChild();
             return new Details(lcurly, rcurly, nextToken, shouldCheckLastRcurly);
         }
 
@@ -451,7 +436,7 @@ public class RightCurlyCheck extends AbstractCheck {
          * @return object containing all details to make a validation
          */
         private static Details getDetailsForIfElse(DetailAST ast) {
-            boolean shouldCheckLastRcurly = false;
+            final boolean shouldCheckLastRcurly;
             final DetailAST lcurly;
             DetailAST nextToken;
             final int tokenType = ast.getType();
@@ -463,6 +448,7 @@ public class RightCurlyCheck extends AbstractCheck {
                     lcurly = ast.getLastChild();
                 }
                 else {
+                    shouldCheckLastRcurly = false;
                     lcurly = nextToken.getPreviousSibling();
                 }
             }
@@ -500,7 +486,7 @@ public class RightCurlyCheck extends AbstractCheck {
                     rcurly = lcurly.getLastChild();
                 }
             }
-            return new Details(lcurly, rcurly, getNextToken(ast), false);
+            return new Details(lcurly, rcurly, getNextToken(ast), true);
         }
 
         /**
@@ -513,7 +499,9 @@ public class RightCurlyCheck extends AbstractCheck {
             final DetailAST lcurly;
             final DetailAST nextToken;
             final int tokenType = ast.getType();
+            final boolean shouldCheckLastRcurly;
             if (tokenType == TokenTypes.LITERAL_DO) {
+                shouldCheckLastRcurly = false;
                 nextToken = ast.findFirstToken(TokenTypes.DO_WHILE);
                 lcurly = ast.findFirstToken(TokenTypes.SLIST);
                 if (lcurly != null) {
@@ -521,6 +509,7 @@ public class RightCurlyCheck extends AbstractCheck {
                 }
             }
             else {
+                shouldCheckLastRcurly = true;
                 lcurly = ast.findFirstToken(TokenTypes.SLIST);
                 if (lcurly != null) {
                     // SLIST could be absent in code like "while(true);"
@@ -528,7 +517,7 @@ public class RightCurlyCheck extends AbstractCheck {
                 }
                 nextToken = getNextToken(ast);
             }
-            return new Details(lcurly, rcurly, nextToken, false);
+            return new Details(lcurly, rcurly, nextToken, shouldCheckLastRcurly);
         }
 
         /**

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheckTest.java
@@ -66,12 +66,16 @@ public class RightCurlyCheckTest extends AbstractModuleTestSupport {
     public void testSame() throws Exception {
         final DefaultConfiguration checkConfig = createModuleConfig(RightCurlyCheck.class);
         checkConfig.addAttribute("option", RightCurlyOption.SAME.toString());
+        checkConfig.addAttribute("tokens", "LITERAL_TRY, LITERAL_CATCH, LITERAL_FINALLY, "
+                + "LITERAL_IF, LITERAL_ELSE, LITERAL_FOR, LITERAL_WHILE, LITERAL_DO");
         final String[] expected = {
             "25:17: " + getCheckMessage(MSG_KEY_LINE_SAME, "}", 17),
             "28:17: " + getCheckMessage(MSG_KEY_LINE_SAME, "}", 17),
             "40:13: " + getCheckMessage(MSG_KEY_LINE_SAME, "}", 13),
             "44:13: " + getCheckMessage(MSG_KEY_LINE_SAME, "}", 13),
             "93:27: " + getCheckMessage(MSG_KEY_LINE_BREAK_BEFORE, "}", 27),
+            "188:9: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 9),
+            "189:53: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 53),
         };
         verify(checkConfig, getPath("InputRightCurlyLeft.java"), expected);
     }
@@ -85,12 +89,32 @@ public class RightCurlyCheckTest extends AbstractModuleTestSupport {
     }
 
     @Test
+    public void testSameDoesNotComplainForNonMultilineConstructs() throws Exception {
+        final DefaultConfiguration checkConfig = createModuleConfig(RightCurlyCheck.class);
+        checkConfig.addAttribute("option", RightCurlyOption.SAME.toString());
+        checkConfig.addAttribute("tokens", "LITERAL_DO, LITERAL_FOR, LITERAL_WHILE, STATIC_INIT,"
+                + "INSTANCE_INIT, CLASS_DEF, METHOD_DEF, CTOR_DEF");
+        final String[] expected = CommonUtil.EMPTY_STRING_ARRAY;
+        verify(checkConfig, getPath("InputRightCurlySame.java"), expected);
+    }
+
+    @Test
     public void testAlone() throws Exception {
         final DefaultConfiguration checkConfig = createModuleConfig(RightCurlyCheck.class);
         checkConfig.addAttribute("option", RightCurlyOption.ALONE.toString());
+        checkConfig.addAttribute("tokens", "LITERAL_TRY, LITERAL_CATCH, LITERAL_FINALLY, "
+                + "LITERAL_IF, LITERAL_ELSE, LITERAL_FOR, LITERAL_WHILE, LITERAL_DO");
         final String[] expected = {
+            "56:13: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 13),
             "93:27: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 27),
             "97:72: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 72),
+            "173:9: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 9),
+            "175:9: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 9),
+            "177:9: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 9),
+            "178:9: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 9),
+            "183:9: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 9),
+            "188:9: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 9),
+            "189:53: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 53),
         };
         verify(checkConfig, getPath("InputRightCurlyLeft.java"), expected);
     }
@@ -115,17 +139,6 @@ public class RightCurlyCheckTest extends AbstractModuleTestSupport {
     }
 
     @Test
-    public void testShouldStartLine() throws Exception {
-        final DefaultConfiguration checkConfig = createModuleConfig(RightCurlyCheck.class);
-        checkConfig.addAttribute("option", RightCurlyOption.ALONE.toString());
-        final String[] expected = {
-            "93:27: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 27),
-            "97:72: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 72),
-        };
-        verify(checkConfig, getPath("InputRightCurlyLeft.java"), expected);
-    }
-
-    @Test
     public void testShouldStartLine2() throws Exception {
         final DefaultConfiguration checkConfig = createModuleConfig(RightCurlyCheck.class);
         checkConfig.addAttribute("option", RightCurlyOption.ALONE_OR_SINGLELINE.toString());
@@ -136,17 +149,6 @@ public class RightCurlyCheckTest extends AbstractModuleTestSupport {
             "122:6: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 6),
             "136:5: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 5),
             "136:6: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 6),
-        };
-        verify(checkConfig, getPath("InputRightCurlyLeft.java"), expected);
-    }
-
-    @Test
-    public void testMethodCtorNamedClassClosingBrace() throws Exception {
-        final DefaultConfiguration checkConfig = createModuleConfig(RightCurlyCheck.class);
-        checkConfig.addAttribute("option", RightCurlyOption.ALONE.toString());
-        final String[] expected = {
-            "93:27: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 27),
-            "97:72: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 72),
         };
         verify(checkConfig, getPath("InputRightCurlyLeft.java"), expected);
     }
@@ -290,6 +292,13 @@ public class RightCurlyCheckTest extends AbstractModuleTestSupport {
             "182:24: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 24),
             "185:24: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 24),
             "188:24: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 24),
+            "194:9: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 9),
+            "196:9: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 9),
+            "198:9: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 9),
+            "199:9: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 9),
+            "204:9: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 9),
+            "209:9: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 9),
+            "210:53: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 53),
         };
         verify(checkConfig, getPath("InputRightCurlyAloneOrSingleline.java"), expected);
     }
@@ -384,6 +393,7 @@ public class RightCurlyCheckTest extends AbstractModuleTestSupport {
         checkConfig.addAttribute("option", RightCurlyOption.ALONE_OR_SINGLELINE.toString());
         final String[] expected = {
             "19:9: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 9),
+            "27:64: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 64),
             "35:15: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 15),
             "37:13: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 13),
         };
@@ -472,6 +482,7 @@ public class RightCurlyCheckTest extends AbstractModuleTestSupport {
             "42:37: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 37),
             "63:43: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 43),
             "67:5: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 5),
+            "71:24: " + getCheckMessage(MSG_KEY_LINE_ALONE, "}", 24),
         };
         verify(checkConfig, getPath(
                 "InputRightCurlyAloneOrSingleLine2.java"), expected);

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/rightcurly/InputRightCurlyAloneOrSingleline.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/rightcurly/InputRightCurlyAloneOrSingleline.java
@@ -187,4 +187,26 @@ public class InputRightCurlyAloneOrSingleline {
         while (true) {
             getClass();} // violation
     }
+
+    public void emptyBlocks() {
+        try {
+            // comment
+        } catch (RuntimeException e) { // violation
+            new Object();
+        } catch (Exception e) { // violation
+            // comment
+        } catch (Throwable e) { // violation
+        } finally { // violation
+            // comment
+        }
+
+        do {
+        } while (true); // violation
+    }
+
+    public void codeAfterLastRightCurly() {
+        while (new Object().equals(new Object())) {
+        }; // violation
+        for (int i = 0; i < 1; i++) { new Object(); }; // violation
+    }
 }

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/rightcurly/InputRightCurlyLeft.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/rightcurly/InputRightCurlyLeft.java
@@ -167,4 +167,25 @@ class ClassWithStaticInitializers
         }
     }
 
+    public void emptyBlocks() {
+        try {
+            // comment
+        } catch (RuntimeException e) { // violation except for SAME
+            new Object();
+        } catch (Exception e) { // violation except for SAME
+            // comment
+        } catch (Throwable e) { // violation except for SAME
+        } finally { // violation except for SAME
+            // comment
+        }
+
+        do {
+        } while (true); // violation except for SAME
+    }
+
+    public void codeAfterLastRightCurly() {
+        while (new Object().equals(new Object())) {
+        }; // violation
+        for (int i = 0; i < 1; i++) { new Object(); }; // violation
+    }
 }

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/rightcurly/InputRightCurlySame.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/rightcurly/InputRightCurlySame.java
@@ -1,0 +1,32 @@
+package com.puppycrawl.tools.checkstyle.checks.blocks.rightcurly;
+
+public class InputRightCurlySame {
+    static {
+    }
+
+    public InputRightCurlySame() {
+        Thread t = new Thread(new Runnable() {
+            {
+            }
+
+            @Override
+            public void run() {
+            }
+        });
+    }
+
+    public void doLoop() {
+        do {
+        } while (true);
+    }
+
+    public void whileLoop() {
+        while (true) {
+        }
+    }
+
+    public void forLoop() {
+        for (; ; ) {
+        }
+    }
+}

